File: app/server/src/p2/session/Session.java
package p2.session;

import p2.media.Catalog;
import p2.media.ImageMethod;
import p2.media.ImageMethod.UnitId;
import p2.media.ImageMethod.Viewport;
import p2.net.Link;
import p2.net.Wire;
import p2.util.Json;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * One viewer, for as long as it is connected: what it is looking at, what it already holds,
 * and what to send next.
 *
 * The rules that matter are here rather than in the transport, so they hold whichever
 * transport is underneath:
 *
 *   - send only units the current view needs, nearest the centre of the screen first;
 *   - never send a unit the client says it holds;
 *   - when the view changes, throw away whatever is still queued for the old one, because
 *     the viewer has moved on and those bytes would arrive already stale;
 *   - stop filling the transport once it says it has enough, so cancelling stays cheap.
 */
public final class Session implements Link.Inbound {

    private final Catalog catalog;
    private final Link link;

    private String imageName;
    private ImageMethod.Served served;
    private int epoch;
    private Viewport viewport;

    /** What the client holds, as it has told us: never send these again unless it drops them. */
    private final Set<UnitId> held = new HashSet<>();
    private List<UnitId> queue = List.of();
    private int queueAt;

    private long unitsSent, bytesSent, unitsCancelled, viewsSeen;
    private long lastStats;
    private boolean wasIdle = true;
    private boolean regather;            // the transport gave work back: work out what is missing

    public Session(Catalog catalog, Link link) {
        this.catalog = catalog;
        this.link = link;
    }

    @Override
    public void message(ByteBuffer message) {
        try {
            if (!Wire.looksValid(message)) {
                fault("not a protocol message");
                return;
            }
            int type = Wire.type(message);
            int messageEpoch = Wire.epoch(message);
            message.position(message.position() + Wire.HEADER);
            switch (type) {
                case Wire.HELLO -> welcome();
                case Wire.OPEN -> open(text(message));
                case Wire.VIEW -> view(messageEpoch, message);
                case Wire.BYE -> link.close();
                default -> fault("unexpected message type " + type);
            }
        } catch (Exception e) {
            fault(String.valueOf(e));
        }
    }

    @Override
    public void closed() {
        served = null;
        queue = List.of();
    }

    // ---------------------------------------------------------------- messages in

    private void welcome() {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("version", Wire.VERSION);
        body.put("method", catalog.method().id());
        send(Wire.text(Wire.WELCOME, epoch, Json.write(body)));
    }

    private void open(String name) throws IOException {
        if (name.equals(imageName) && served != null) {
            chart();                    // already open: they only missed the answer
            return;
        }
        if (!catalog.has(name)) {
            fault("no such image: " + name);
            return;
        }
        if (!catalog.method().isPrepared(catalog.preparedDir(name))) {
            fault(name + " is not prepared yet - prepare it on the server page first");
            return;
        }
        imageName = name;
        served = catalog.served(name);
        held.clear();
        queue = List.of();
        queueAt = 0;
        unitsSent = bytesSent = unitsCancelled = viewsSeen = 0;
        // The epoch is deliberately left alone: it counts the viewer's moves for as long as
        // the session lasts, and both ends read a lower one as an older view.

        chart();
    }

    /** What this image looks like: its shape, its method, how it is cut up. */
    private void chart() {
        ImageMethod.Meta meta = served.meta();
        Map<String, Object> chart = new LinkedHashMap<>();
        chart.put("image", imageName);
        chart.put("method", catalog.method().id());
        chart.put("contentType", served.unitContentType());
        chart.put("width", meta.width());
        chart.put("height", meta.height());
        chart.put("unitSize", meta.unitSize());
        chart.put("ratio", meta.ratio());
        chart.put("maxLevel", meta.maxLevel());
        chart.put("units", meta.units());
        chart.put("bytes", meta.bytes());
        send(Wire.text(Wire.CHART, epoch, Json.write(chart)));
    }

    /**
     * A new view. Payload: centre x and y and scale as doubles, the screen size, then the
     * units the client has dropped since last time (it is bounded by its memory budget, and
     * we must know, or we would never send them again).
     */
    private void view(int viewEpoch, ByteBuffer b) throws IOException {
        if (served == null) {
            fault("open an image first");
            return;
        }
        double cx = b.getDouble(), cy = b.getDouble(), scale = b.getDouble();
        int screenW = b.getShort() & 0xffff, screenH = b.getShort() & 0xffff;
        int dropped = b.getShort() & 0xffff;
        for (int i = 0; i < dropped; i++) {
            held.remove(new UnitId(b.getShort() & 0xffff, b.getInt(), b.getInt()));
        }
        if (viewEpoch < epoch) return;                  // an older view overtook a newer one
        viewsSeen++;
        epoch = viewEpoch;
        viewport = new Viewport(cx, cy, scale, screenW, screenH);

        List<UnitId> wanted = served.unitsFor(viewport);
        List<UnitId> next = new ArrayList<>(wanted.size());
        for (UnitId id : wanted) {
            if (!held.contains(id)) next.add(id);
        }
        unitsCancelled += Math.max(0, queue.size() - queueAt);   // whatever the old view still wanted
        queue = next;
        queueAt = 0;
        pump();
    }

    // ---------------------------------------------------------------- messages out

    /** Hand the transport as much as it will take, in priority order, for the current view. */
    private void pump() throws IOException {
        while (queueAt < queue.size() && link.writable()) {
            UnitId id = queue.get(queueAt++);
            if (held.contains(id)) continue;
            byte[] payload = served.bytes(id);
            if (!link.send(Wire.unit(epoch, id.level(), id.x(), id.y(), payload).array())) break;
            held.add(id);
            unitsSent++;
            bytesSent += payload.length;
        }
        stats();
    }

    /**
     * The transport gave up on a unit. Strike it from the ledger, and if the view still wants
     * it, queue it again.
     *
     * Without this the ledger lies. A unit is written into it the moment it is handed over,
     * because the transport is the thing that knows how to deliver it; when the transport then
     * abandons it - the viewer moved, the deadline passed - the session would go on believing
     * the viewer had a tile it was never sent, and would never offer it again. The hole stays
     * on the screen for as long as the viewer keeps looking at it, which is exactly the case
     * where it is most obvious.
     */
    @Override
    public void abandoned(ByteBuffer message) {
        if (!Wire.looksValid(message) || Wire.type(message) != Wire.UNIT) return;
        int at = message.position() + Wire.HEADER;
        UnitId id = new UnitId(message.getShort(at) & 0xffff,
                message.getInt(at + 2), message.getInt(at + 6));
        if (!held.remove(id)) return;
        unitsSent--;
        bytesSent -= message.remaining() - Wire.HEADER - Wire.UNIT_HEADER;
        // Not queued here and now: this is called from inside the sending loop, which is
        // walking the very queue that would be replaced. The next turn of the loop picks it up.
        regather = true;
    }

    /** Work out again what the current view is missing, after units were given back. */
    private void regather() {
        regather = false;
        if (served == null || viewport == null) return;
        List<UnitId> next = new ArrayList<>();
        for (UnitId id : served.unitsFor(viewport)) {
            if (!held.contains(id)) next.add(id);
        }
        queue = next;
        queueAt = 0;
    }

    /** Called by the transport when it drains, so a big view keeps flowing. */
    @Override
    public void drained() {
        try {
            if (regather) regather();
            pump();
        } catch (IOException e) {
            fault(String.valueOf(e));
        }
    }

    /**
     * Statistics for the panel, a few times a second at most, plus one the moment the queue
     * empties so the last numbers are the true ones.
     *
     * The throttle has to hold even when there is nothing left to send. Over our own protocol
     * this is called every time the client reports, a hundred times a second, and each of
     * these is a message the sender then has to deliver: for one image of twelve tiles the
     * wire was carrying seven hundred units, almost all of them statistics about carrying
     * statistics.
     */
    private void stats() {
        long now = System.currentTimeMillis();
        boolean idle = queueAt >= queue.size();
        boolean justFinished = idle && !wasIdle;
        wasIdle = idle;
        if (now - lastStats < 200 && !justFinished) return;
        lastStats = now;
        Map<String, Object> s = new LinkedHashMap<>();
        s.put("epoch", epoch);
        s.put("image", imageName);
        s.put("queued", Math.max(0, queue.size() - queueAt));
        s.put("held", held.size());
        s.put("unitsSent", unitsSent);
        s.put("bytesSent", bytesSent);
        s.put("unitsCancelled", unitsCancelled);
        s.put("views", viewsSeen);
        s.put("linkBytes", link.bytesSent());
        s.put("linkMessages", link.messagesSent());
        s.putAll(link.extra());
        send(Wire.text(Wire.STATS, epoch, Json.write(s)));
    }

    private void fault(String text) {
        send(Wire.text(Wire.FAULT, epoch, text));
    }

    private void send(byte[] message) {
        link.send(message);
    }

    private static String text(ByteBuffer b) {
        byte[] bytes = new byte[b.remaining()];
        b.get(bytes);
        return new String(bytes, StandardCharsets.UTF_8);
    }
}

File: splat-pyramid/server/session.ts
// One viewer's session: what it is looking at, what it already holds, what to send next.

import { readFileSync } from "node:fs";
import {
  Type, KIND_SPLAT, decodeReport, decodeView, encode, unitKey,
  type Message, type UnitId, type View,
} from "../shared/wire.ts";
import { confetti, readSpx, tileParts } from "../shared/units.ts";
import type { PreparedImage } from "./image.ts";

/**
 * Packets of units, prepared once and shared by every session: the second viewer of a unit
 * costs no disk read and no packetising. Least recently used units go first over the budget.
 */
export class PacketCache {
  private units = new Map<string, Buffer[]>();
  private bytes = 0;
  hits = 0;
  misses = 0;
  private budget: number;

  constructor(budget = 128 * 2 ** 20) {
    this.budget = budget;
  }

  packets(image: PreparedImage, u: UnitId): Buffer[] {
    const key = `${image.chart.name}/${unitKey(u)}`;
    const found = this.units.get(key);
    if (found) {
      this.units.delete(key);
      this.units.set(key, found);
      this.hits++;
      return found;
    }
    this.misses++;
    const made = build(image, u);
    this.units.set(key, made);
    this.bytes += made.reduce((s, p) => s + p.length, 0);
    for (const [k, v] of this.units) {
      if (this.bytes <= this.budget) break;
      this.units.delete(k);
      this.bytes -= v.reduce((s, p) => s + p.length, 0);
    }
    return made;
  }
}

function build(image: PreparedImage, u: UnitId): Buffer[] {
  if (u.kind === KIND_SPLAT) return confetti(u.level, u.x, u.y, readSpx(image.splatPath(u.level, u.x, u.y)));
  const file = image.tileFile(u.level, u.x, u.y);
  if (!file) return [];
  return tileParts(u.level, u.x, u.y, file.format, readFileSync(file.path));
}

export interface Peer {
  address: string;
  port: number;
}

export class Session {
  epoch = 0;
  image: PreparedImage | null = null;
  view: View | null = null;
  lastHeard = Date.now();
  private held = new Set<string>();      // units the viewer has (or is being sent) in full
  private queue: UnitId[] = [];
  private current: Buffer[] = [];        // packets of the unit being sent
  private currentAt = 0;
  private currentType: typeof Type.CONFETTI | typeof Type.TILEPART = Type.CONFETTI;
  // counters, for STATS
  packetsSent = 0;
  bytesSent = 0;
  unitsSent = 0;
  viewsSeen = 0;
  cancelled = 0;
  reported = { packets: 0, bytes: 0 };

  readonly peer: Peer;
  readonly send: (msg: Buffer) => void;
  private images: Map<string, PreparedImage>;
  private cache: PacketCache;

  constructor(peer: Peer, images: Map<string, PreparedImage>, cache: PacketCache,
              send: (msg: Buffer) => void) {
    this.peer = peer;
    this.images = images;
    this.cache = cache;
    this.send = send;
  }

  onMessage(m: Message): void {
    this.lastHeard = Date.now();
    switch (m.type) {
      case Type.HELLO:
        this.reset();
        this.send(encode(Type.WELCOME, 0));
        break;
      case Type.OPEN: {
        const name = m.payload.toString("utf8");
        const image = this.images.get(name);
        if (!image) {
          this.send(encode(Type.FAULT, m.epoch, Buffer.from(
            `no image "${name}"; there is: ${[...this.images.keys()].join(", ")}`)));
          return;
        }
        this.reset();
        this.image = image;
        this.send(encode(Type.CHART, m.epoch, Buffer.from(JSON.stringify(image.chart))));
        break;
      }
      case Type.VIEW: {
        if (!this.image) return;
        const v = decodeView(m.payload);
        for (const u of v.dropped) this.held.delete(unitKey(u));
        if (m.epoch < this.epoch) return;          // an older view overtook a newer one
        this.epoch = m.epoch;
        this.view = v;
        this.viewsSeen++;
        const wanted = this.image.unitsFor(v).filter((u) => !this.held.has(unitKey(u)));
        this.cancelled += Math.max(0, this.queue.length);
        this.queue = wanted;
        break;
      }
      case Type.REPORT:
        this.reported = decodeReport(m.payload);
        break;
      case Type.BYE:
        this.reset();
        break;
    }
  }

  private reset(): void {
    this.image = null;
    this.view = null;
    this.held.clear();
    this.queue = [];
    this.current = [];
    this.currentAt = 0;
  }

  /** Send up to `budget` bytes, most important first. Returns the bytes sent. */
  pump(budget: number): number {
    let spent = 0;
    while (spent < budget) {
      if (this.currentAt >= this.current.length) {
        const next = this.queue.shift();
        if (!next || !this.image) break;
        const key = unitKey(next);
        if (this.held.has(key)) continue;
        this.current = this.cache.packets(this.image, next);
        this.currentAt = 0;
        this.currentType = next.kind === KIND_SPLAT ? Type.CONFETTI : Type.TILEPART;
        this.held.add(key);
        this.unitsSent++;
      }
      const payload = this.current[this.currentAt++];
      const msg = encode(this.currentType, this.epoch, payload);
      this.send(msg);
      spent += msg.length;
      this.packetsSent++;
      this.bytesSent += msg.length;
    }
    return spent;
  }

  get idle(): boolean {
    return this.currentAt >= this.current.length && this.queue.length === 0;
  }

  stats(rate: number): Record<string, unknown> {
    return { epoch: this.epoch, views: this.viewsSeen, queued: this.queue.length,
             unitsSent: this.unitsSent, packetsSent: this.packetsSent, bytesSent: this.bytesSent,
             cancelled: this.cancelled, reported: this.reported, rate };
  }
}
